### DISCUSSION TITLE
What configuration for RUC7258 chip?

### DISCUSSION THREAD

**Discussion by Kalrkloss** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#501] [Q&A]
The RUC7258 is in the list of supported chips, but nowhere in the sourcecode.
What driver option do I actually have to choose for a RUC7258 based 1/16 panel (64x32)?

**Comment by board707** [comment]
The ruc7258 is not a driver, it is just a multiplexor. To make conclusion, is your panel compatible with the library or not, we need to know the type of the another chip on the panel.

Addition: 7258 multiplexor doesn't require any special configuration

**Reply by Kalrkloss** [reply]
The panel has (amongst some 74HC245) RUC7258 and  RUL6024.
So, how to configure the code?

**Reply by Kalrkloss** [reply]
I can't find a datasheet for the RUL6024, is it possible that it's a clone of the MBI6024?

If it is compatible, then the RUC7258 is a high-side driver/multiplexer that controls the rows, and the RUL6024 is a low-side  constant current LED driver that controls the columns. Is that right?
The RUC7258 is then a simple multiplexer that needs only TTL signals, but no further configuration, while the RUL6024 needs special code to configure it and receive the pixel data via SPI.

**Reply by daveythacher** [reply]
There is usually three groups of ICs. The quantity gives them away. Low side is usually something like columns / 16 * 6. (However it may be higher, see below.)

There is a very small number of octal bi-directional bus transceivers which are the 74HC245 (CMOS) or 74HCT245 (TTL). These are used to boost the signal for the fan out inside the panel. Note these are passive and will eventually lead to corruption. All signal pass through these.

The next most common IC is the high side driver. They use a couple different ones here. On some it is dual P-MOSFETs. On others it is P-MOSFET arrays like the RUC7258. Only A, B, C D, and E go to these. There are several types Direct, Decoder, Shift Register and hybrids.

The most common IC is the constant current low side driver. These are the ones which usually have all the features implemented on them. These have R0, G0, B0, R1, G1, B1, CLK, LAT,  and OE go to them. Note R0, G0, B0, etc. are in their on shift register chain. Note some low side drivers require an initialization sequence before they will display any information. These are not likely supported by many code bases.

Low side can implement different topologies depending on the needs of the panel. Most common in outdoor panels which need to be slightly brighter. Z-strip mappers are sometimes required accordingly. In these topologies the panels may simulate a higher multiplex than HUB75 allows. This increases the chain length while giving more scan lines.

**Reply by board707** [reply]
> I can't find a datasheet for the RUL6024, is it possible that it's a clone of the MBI6024?

Yes, it is possible.
If so, then this panel is not supported by this library and most likely not supported by any library at all

**Reply by board707** [reply]
> This increases the chain length while giving more scan lines.

Did you want to write "LESS scan lines" ? because Z-strip patterns are usually used to reduce the number of scans on a panel

**Reply by daveythacher** [reply]
Panels from Sparkfun and Adafruit are almost always compatible. There are other sellers also DFRobot and WiredWatts that are also targeting compatibility. Digikey carries panels from them also. 

I got really luck once with an order from Adafruit, but it was likely just luck, I got MBI5124 low side, rather than the run of the mill stuff they normally hand out. DFRobot sells low refresh panels. (They ghost like crazy above 240/250Hz.) I want to say the panels I got from Sparkfun were pretty good, but I have only ordered once from them and it was a different configuration. Never ordered the panels from WireWatts to my knowledge.

**Reply by daveythacher** [reply]
Quote reply is broken. Github you slipping. (Why they hire these people, they work fine then boom basic features don't work.)

> Did you want to write "LESS scan lines" ? because Z-strip patterns are usually used to reduce the number of scans on a panel

HUB75 has two scan lines (aka two lines will be on all the time), with Z-strip they usually increase this to four scan lines (aka four lines will be on all the time). There is not enough data pins for four so they increase the columns and map the columns to two rows per half the display. This is a tight fit for routing and board space, so they had to get cleaver. Thus Z-strip came to be. Basically you have two 8x32 panels mixed together in a 16x32 panel. (Under the covers it is a 8x64 pretending to be 16x32.)

So the S number is less. However the active lines is higher. Scan technically like three or four things.

**Reply by board707** [reply]
> Quote reply is broken.

I just copy quotes to reply by hands :)

>So the S number is less. However the active lines is higher.

Thanks, I understood

**Reply by Kalrkloss** [reply]
The thing is, I can almost drive this panel with a Color light 5a-75b board and Ledvision.
"Almost" meaning that only the upper half of the display works and is somewhat garbled.

**Reply by daveythacher** [reply]
I looked into this and it works for the most part. However the Ethernet packets are supposed to be very consistent in timing. The display glitches when this is not true. I found that using a PIC32MZ with a dedicated 1G switch fixed the issue.

What I did was carefully write firmware in the PIC32MZ to move data from USB HS (single bulk endpoint) to 100Mb Ethernet. The switch then proxies this to 1Gb Ethernet for the receiver card. I was able to get the transfers very consistent with this. The cost is somewhat high and my USB logic was a total hack. (I eventually stopped work on it due to legal concerns.)

Raspberry Pi was capable of driving a decent amount of pixels directly for non GCLK (hardware PWM) panels. This is Pi first gen, but second gen was significantly faster.

**Reply by daveythacher** [reply]
Which 5A-7B firmware did you use? (PWM or the standard one)

**Reply by Kalrkloss** [reply]
I didn't even know that there are different ones. I used the one it came with.

**Reply by daveythacher** [reply]
If that worked then you are not likely using a hardware PWM panel. (That means support would be easier to add to this project. Assuming it can be done legally.)

You may need to adjust your settings. You need to define both the cabinet and the panel size. It assumes all panels are the same within the cabinet. The eight connectors are stacked by row. (You may be able to change it, but I just did this in software on PC/Pi application.) You need to know the scan, columns and pixel locations. There is a walk thru which helps you out. The netcard function is a setup tool which can be abused. I forget if it gives you snow or just repeats the image. There is a number of knobs and switches, but most are for random cases. (From my tinkering at least, my guess is they help with compatibility electrically.)

Note for high refresh you need to use something with PWM rather than BCM. I do not recall which ones they use, but I believe it is BCM. PWM uses more memory and computational power. I think they use a 32 or 64 bit SDRAM/DDR interface. They can do PWM but they would have to scale it down. For the most part standard firmware likely is limited on refresh, however an alternative which reduces color depth may be able to fix this.

**Comment by daveythacher** [comment]
It is supported by the code base. It is an optimized high side driver. (It handles the rows.) It likely has some kind of push/pull driver for managing upper ghosting, which is pretty much required for high refresh rates. This is passive to the code base. (Any optimizations that this brings passively are not fully used by this code base.) This code base does not support super high refreshes like most BCM code bases.

Just enjoy the improved performance.

**Reply by board707** [reply]
@daveythacher 
Is this an output from Chatgpt?
A lot of empty phrases that have nothing to do with the topic.

**Reply by daveythacher** [reply]
I actually did play with hardware. (Connect the dots.) I actually did a little more than play...

**Reply by board707** [reply]
I hope we see something concrete other than words
